### PR TITLE
Chapter 3

### PR DESCRIPTION
**Pull request by Moonforeva**
Ready

### CODE REVIEW

**Comment by Moonforeva** [comment]
Ready

File: .gitignore [hunk 1]
```diff
@@ -1,56 +1,39 @@
...
+
+/public/packs
+/public/packs-test
+/node_modules
+/yarn-error.log
+yarn-debug.log*
+.yarn-integrity
+/config/credentials.yml.enc
+/config/database.yml
+/test
```
Comment (saobv-2223) [inline .gitignore:39]: add end line, please

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * compiled file so the styles you add here take precedence over styles defined in any other CSS/SCSS
+ * files in this directory. Styles in this file should be added after the last require_* statement.
+ * It is generally better to create a new file per style scope.
+ *
+ *= require_tree .
+ *= require_self
+ */
+
+ @import "bootstrap-sprockets";
+ @import "bootstrap";
```
Comment (saobv-2223) [inline app/assets/stylesheets/application.scss:18]:
```suggestion
 @import 'bootstrap-sprockets';
 @import 'bootstrap';
 
```

File: app/assets/stylesheets/custom.scss [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+section {
+    overflow: auto;
+}
+
+textarea {
+    resize: vertical;
+}
+
+.center {
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss:14]: indenting only 2 spaces
add an empty line below the last line

File: app/assets/stylesheets/static_pages.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the StaticPages controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/static_pages.scss:3]: remove file that never used

File: app/controllers/static_pages_controller.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class StaticPagesController < ApplicationController
+  def home
+  end
```
Comment (saobv-2223) [inline app/controllers/static_pages_controller.rb:3]:
```suggestion
  def home;  end
```

File: app/controllers/static_pages_controller.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+
+  def help
+  end
```
Comment (saobv-2223) [inline app/controllers/static_pages_controller.rb:6]:
```suggestion
  def help; end
```

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+module ApplicationHelper
+    def full_title page_title = ""
+        base_title = "Ruby on Rails Tutorial Sample App"
+        page_title.empty? ? base_title : "#{page_title} | #{base_title}"
+    end
```
Comment (saobv-2223) [inline app/helpers/application_helper.rb:5]:
```suggestion
    def full_title page_title
        base_title = "Ruby on Rails Tutorial Sample App"
        page_title.blank? ? base_title : "#{page_title} | #{base_title}"
    end
```

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+      <%= link_to "sample app", "#", id: "logo" %>
+      <nav>
+        <ul class="nav navbar-nav navbar-right">
+          <li><%= link_to "Home", "#" %></li>
+          <li><%= link_to "Help", "#" %></li>
+          <li><%= link_to "Log in", "#" %></li>
+        </ul>
+      </nav>
+    </div>
+  </header>
```
Comment (saobv-2223) [inline app/views/layouts/_header.html.erb:12]: indenting only 2 spaces, add an empty line below line 12

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title><%= full_title(yield(:title))%></title>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb:4]:
```suggestion
    <title><%= full_title yield(:title) %></title>
```

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    <%= javascript_pack_tag "application", "data-turbolinks-track": "reload" %>
+  </head>
+
+  <body>
+  <%= render "layouts/header"%>
+  <div class="container">
+    <%= yield %>
+  </div>
+
+</body>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb:18]:
```suggestion
  <body>
    <%= render "layouts/header"%>
    <div class="container">
      <%= yield %>
    </div>
  </body>
```

File: app/views/static_pages/help.html.erb [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+<h1>Help</h1>
+<p>
+    Get help on the Ruby on Rails Tutorial at the
+    <a href="http://www.railstutorial.org/help">Rails Tutorial
+help page</a>.
+    To get help on this sample app, see the
+    <a href="http://www.railstutorial.org/book"><em>Ruby on Rails
+Tutorial</em>
+    book</a>.
```
Comment (saobv-2223) [inline app/views/static_pages/help.html.erb:11]: using link_to instead of a tag

File: app/views/static_pages/help.html.erb [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+</p>
```
Comment (saobv-2223) [inline app/views/static_pages/help.html.erb:12]: add end line, please

File: app/views/static_pages/home.html.erb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+<%= provide :title, "Home" %>
+
+<h1>Sample App</h1>
+<p>
+    This is the home page for the
+<a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
+    sample application.
+</p>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb:8]: same as above

**Comment by Moonforeva** [comment]
Ready

File: app/assets/stylesheets/application.scss [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+ * files in this directory. Styles in this file should be added after the last require_* statement.
+ * It is generally better to create a new file per style scope.
+ *
+ *= require_tree .
+ *= require_self
+ */
+
+ @import 'bootstrap-sprockets';
+ @import 'bootstrap';
+
```
Comment (saobv-2223) [inline app/assets/stylesheets/application.scss:19]: image

File: app/assets/stylesheets/custom.scss [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+
+
+/* header */
+
+#logo {
+  float: left;
+  margin-right: 10px;
+  font-size: 1.7em;
+  color: #fff;
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss:30]: image

File: app/assets/stylesheets/static_pages.scss (hunk 4 above)
Comment (saobv-2223) [reply app/assets/stylesheets/static_pages.scss:3]: image

File: app/helpers/application_helper.rb [hunk 16]
```diff
@@ -0,0 +1,6 @@
+module ApplicationHelper
+    def full_title page_title
+        base_title = "Ruby on Rails Tutorial Sample App"
+        page_title.blank? ? base_title : "#{page_title} | #{base_title}"
+    end
```
Comment (saobv-2223) [inline app/helpers/application_helper.rb:5]: image

File: app/javascript/packs/application.js [hunk 17]
```diff
@@ -0,0 +1,9 @@
+
+
+require('@rails/ujs').start()
+require('turbolinks').start()
+require('@rails/activestorage').start()
+require('channels')
+
+
+
```
Comment (saobv-2223) [inline app/javascript/packs/application.js:9]:
```suggestion
```

File: app/views/layouts/application.html.erb (hunk 9 above)
Comment (saobv-2223) [reply app/views/layouts/application.html.erb:4]: fix it, please

**Review by saobv-2223 — COMMENTED** [review]
You PR is having 2 commits, merged to 1 commit, please

**Comment by Moonforeva** [comment]
Ready

File: .gitignore [hunk 18]
```diff
@@ -18,11 +18,14 @@
...
 !/tmp/.keep
 
+<<<<<<< HEAD
 # Ignore pidfiles, but keep the directory.
 /tmp/pids/*
 !/tmp/pids/
 !/tmp/pids/.keep
 
+=======
+>>>>>>> 6654c04... Init Project
```
Comment (manhbnt-0960) [inline .gitignore:28]: please fix conflict

File: Gemfile [hunk 19]
```diff
@@ -1,6 +1,7 @@
 source "https://rubygems.org"
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
+<<<<<<< HEAD
```
Comment (manhbnt-0960) [inline Gemfile:4]: semilar

File: Gemfile [hunk 20]
```diff
@@ -17,30 +18,63 @@ gem "webpacker", "~> 4.0"
...
+  gem "capybara", ">= 2.15"
+  gem "selenium-webdriver"
+  gem "webdrivers"
+end
+
+group :development, :test do
+  gem "rubocop", "~> 0.74.0", require: false
+  gem "rubocop-rails", "~> 2.3.2", require: false
+end
+>>>>>>> 6654c04... Init Project
```
Comment (manhbnt-0960) [inline Gemfile:78]: semilar

File: app/helpers/static_pages_helper.rb [hunk 21]
```diff
@@ -0,0 +1,2 @@
+module StaticPagesHelper
```
Comment (manhbnt-0960) [inline app/helpers/static_pages_helper.rb:1]: delete this empty file

File: config/environments/test.rb [hunk 22]
```diff
@@ -7,7 +7,10 @@
   # Settings specified here will take precedence over those in config/application.rb.
 
   config.cache_classes = false
+<<<<<<< HEAD
```
Comment (manhbnt-0960) [inline config/environments/test.rb:10]: remove <<<< HEAD

**Review by manhbnt-0960 — COMMENTED** [review]
you must run "bundle install" after edit Gemfile

**Comment by Moonforeva** [comment]
Ready

File: app/assets/stylesheets/application.scss (hunk 14 above)
Comment (saobv-2223) [reply app/assets/stylesheets/application.scss:19]: fix it, please

File: app/assets/stylesheets/color.scss [hunk 23]
```diff
@@ -0,0 +1,17 @@
+/* header */
+
+#logo {
+  float: left;
+  margin-right: 10px;
+  font-size: 1.7em;
+  color: #fff;
```
Comment (saobv-2223) [inline app/assets/stylesheets/color.scss:7]: image

File: app/assets/stylesheets/custom.scss [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+.center {
+  text-align: center;
+}
+
+.center h1 {
+  margin-bottom: 10px;
+}
+
+@import 'color';
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss:22]: move import to top of file

File: bin/bundle [hunk 25]
```diff
@@ -1,4 +1,8 @@
+<<<<<<< HEAD
 #!/usr/bin/env ruby2.7
+=======
+#!/usr/bin/env ruby
+>>>>>>> 6654c04... Init Project
```
Comment (saobv-2223) [inline bin/bundle:5]: resolve conflict

File: config/locales/en.yml [hunk 26]
```diff
@@ -31,3 +31,8 @@
 
 en:
   hello: "Hello world"
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline config/locales/en.yml:34]: resolve conflict

File: config/routes.rb [hunk 27]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline config/routes.rb:2]: resolve conflict and using "

File: package.json [hunk 28]
```diff
@@ -5,7 +5,11 @@
     "@rails/actioncable": "^6.0.0",
     "@rails/activestorage": "^6.0.0",
     "@rails/ujs": "^6.0.0",
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline package.json:8]: resolve conflict

**Review by saobv-2223 — COMMENTED** [review]
conflict

**Comment by Moonforeva** [comment]
Ready

File: app/views/static_pages/home.html.erb [hunk 29]
```diff
@@ -0,0 +1,8 @@
+<%= provide :title, "Home" %>
+
+<h1>Sample app</h1>
```
Comment (ngocvt-0484) [inline app/views/static_pages/home.html.erb:3]: please learn about i18n and use it here
https://github.com/ruby-i18n/i18n

**Comment by Moonforeva** [comment]
Ready

File: app/assets/stylesheets/application.scss (hunk 14 above)
Comment (saobv-2223) [reply app/assets/stylesheets/application.scss:18]: fix it, please

File: app/assets/stylesheets/custom.scss [hunk 30]
```diff
@@ -0,0 +1,40 @@
+@import 'color.scss';
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss:1]:
```suggestion
@import 'color';
```

File: app/views/static_pages/help.html.erb [hunk 31]
```diff
@@ -0,0 +1,9 @@
+<%= t( provide :title, "Help" ) %>
```
Comment (saobv-2223) [inline app/views/static_pages/help.html.erb:1]: remove ()

**Comment by Moonforeva** [comment]
Ready

**Comment by manhbnt-0960** [comment]
merged

**Pull request merged**